app/actions/v3/service_instance_update_managed.rb

```diff
@@ -0,0 +1,372 @@
+          save_incomplete_instance(service_instance, details)
+        else
+          complete_instance_and_save(service_instance, details)
+        end
+      rescue => e
+        save_failed_state(service_instance, e)
+
+        raise e
+      end
+
+      def poll
+        client = VCAP::Services::ServiceClientProvider.provide(instance: service_instance)
+        details = client.fetch_service_instance_last_operation(service_instance, user_guid: @user_audit_info.user_guid)
+
+        case details[:last_operation][:state]
+        when 'succeeded'
+          # TODO: If instance retrievable update dashboard
+          # params = client.fetch_service_instance(instance, user_guid: @user_audit_info.user_guid)
+          complete_instance_and_save(service_instance, parse_response(details))
+          return PollingFinished
+        when 'in progress'
+          save_last_operation(service_instance, details[:last_operation])
+          ContinuePolling.call(details[:retry_after])
+        when 'failed'
+          save_last_operation(service_instance, details[:last_operation])
+          raise LastOperationFailedState.new(details[:last_operation][:description])
+        end
+      rescue LastOperationFailedState => e
+        raise e
+      rescue => e
+        save_failed_state(service_instance, e)
+        raise e
+      end
+
+      private
+
+      attr_reader :service_instance, :message
+
+      def event_repository
+        Repositories::ServiceEventRepository.new(@user_audit_info)
+      end
+
+      def complete_instance_and_save(instance, broker_response)
+        updates = message.updates.tap do |u|
+          u[:service_plan_guid] = service_plan.guid
+          u[:dashboard_url] = broker_response[:dashboard_url] if broker_response.key?(:dashboard_url)
+          u[:maintenance_info] = maintenance_info if maintenance_info_updated?
+        end
+
+        ManagedServiceInstance.db.transaction do
+          service_instance.save_with_new_operation(
+            updates,
+            broker_response[:last_operation] || {}
+          )
+          MetadataUpdate.update(service_instance, message)
+        end
+
+        event_repository.record_service_instance_event(:update, instance, @audit_hash)
+      end
+
+      def save_incomplete_instance(instance, broker_response)
+        attributes_to_update = {}
+        attributes_to_update[:dashboard_url] = broker_response[:dashboard_url] if broker_response.key?(:dashboard_url)
+
+        ManagedServiceInstance.db.transaction do
+          instance.lock!
+          instance.last_operation.lock! if instance.last_operation
+          instance.save_with_new_operation(
+            attributes_to_update,
+            broker_response[:last_operation] || {}
+          )
+        end
+
+        event_repository.record_service_instance_event(:start_update, instance, @audit_hash)
+      end
+
+      def save_failed_state(instance, e)
+        instance.save_with_new_operation(
+          {},
```

Comment:
Not sure I follow why sometimes we lock last_operation and sometimes we don't. In particular we used to lock when we saved the failed state. So maybe it's worth reviewing it.